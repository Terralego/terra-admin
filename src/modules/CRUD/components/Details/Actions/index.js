import { withNamespaces } from 'react-i18next';
import { withRouter } from 'react-router-dom';
import { connectAuthProvider } from '@terralego/core/modules/Auth';

import { connectCRUDProvider } from '../../../services/CRUDProvider';
import { getLayer } from '../../../services/CRUD';

import Actions from './Actions';

export default withRouter(
  connectAuthProvider(({
    authenticated,
    user,
  }, { displayUpdate, displayDelete }) => {
    const permissions = authenticated ? user.permissions : [];
    return {
      displayDelete: displayDelete && permissions.includes('geostore.delete_feature'),
      displayUpdate: displayUpdate && permissions.includes('geostore.change_feature'),
    };
  })(
    connectCRUDProvider(({
      getSettings,
      settings,
      deleteFeature,
    }, {
      match: { params: { layer } },
    }) => ({
      getSettings,
      deleteFeature,
      layer: getLayer(settings, layer),
    }))(
      withNamespaces()(Actions),
    ),
  ),
);
