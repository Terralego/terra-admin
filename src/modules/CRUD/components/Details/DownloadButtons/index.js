import { withNamespaces } from 'react-i18next';

import compose from '../../../../../utils/compose';
import DownloadButtons from './DownloadButtons';

export default compose(
  withNamespaces(),
)(DownloadButtons);
