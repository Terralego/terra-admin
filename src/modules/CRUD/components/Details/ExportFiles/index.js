import ExportFiles from './ExportFiles';

export default ExportFiles;
