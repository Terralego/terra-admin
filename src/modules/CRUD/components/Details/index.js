import { withRouter } from 'react-router-dom';
import { withTranslation } from 'react-i18next';

import { connectCRUDProvider } from '../../services/CRUDProvider';
import { getView } from '../../services/CRUD';

import Details from './Details';

export default withRouter(
  connectCRUDProvider(({
    settings,
    fetchFeature,
    feature,
    errors,
  }, {
    match: { params: { layer, id } },
  }) => {
    const {
      error = {},
      error: { message = '' } = {},
    } = errors.feature.find(({ featureId }) => featureId === id) || {};
    return {
      fetchFeature,
      feature: feature[id] || {},
      view: getView(settings, layer),
      hasError: !!error,
      errorMessage: message,
    };
  })(
    withTranslation()(Details),
  ),
);
