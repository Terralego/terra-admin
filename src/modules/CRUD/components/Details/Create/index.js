import { withRouter } from 'react-router-dom';
import { withTranslation } from 'react-i18next';
import { connectAuthProvider } from '@terralego/core/modules/Auth';

import { connectAppProvider } from '../../../../../components/AppProvider';
import { connectCRUDProvider } from '../../../services/CRUDProvider';
import { getView } from '../../../services/CRUD';
import compose from '../../../../../utils/compose';

import Create from './Create';

const appProviderGetter = ({
  env: { modules: { CRUD: { settings } } },
}) => ({
  settingsEndpoint: settings,
});

const authProviderGetter = ({
  authenticated,
  user,
}) => {
  const permissions = authenticated ? user.permissions : [];
  return {
    displayAddFeature: permissions.includes('can_add_feature'),
    displayChangeFeature: permissions.includes('can_change_feature'),
  };
};

const CRUDPRoviderGetter = ({
  errors: {
    feature,
  },
  getFeaturesList,
  settings,
  saveFeature,
}, {
  match: { params: { layer } },
}) => ({
  featureError: feature[0],
  getFeaturesList,
  saveFeature,
  view: getView(settings, layer),
});

export default compose(
  withRouter,
  connectAppProvider(appProviderGetter),
  connectAuthProvider(authProviderGetter),
  connectCRUDProvider(CRUDPRoviderGetter),
  withTranslation(),
)(Create);
