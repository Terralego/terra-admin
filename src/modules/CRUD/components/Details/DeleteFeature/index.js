import { withTranslation } from 'react-i18next';
import { withRouter } from 'react-router-dom';
import { connectAuthProvider } from '@terralego/core/modules/Auth';

import { connectAppProvider } from '../../../../../components/AppProvider';
import { connectCRUDProvider } from '../../../services/CRUDProvider';
import { getView } from '../../../services/CRUD';
import compose from '../../../../../utils/compose';

import DeleteFeature from './DeleteFeature';

const appProvider = ({
  env: { modules: { CRUD: { settings } } },
}) => ({
  settingsEndpoint: settings,
});

const authProvider = ({
  authenticated,
  user,
}) => {
  const permissions = authenticated ? user.permissions : [];
  return {
    displayDelete: permissions.includes('can_delete_feature'),
  };
};

const CRUDPRovider = ({
  feature,
  getSettings,
  settings,
  deleteFeature,
}, {
  match: { params: { id, layer } },
}) => ({
  feature: feature[id],
  getSettings,
  deleteFeature,
  view: getView(settings, layer),
});

export default compose(
  withRouter,
  connectAppProvider(appProvider),
  connectAuthProvider(authProvider),
  connectCRUDProvider(CRUDPRovider),
  withTranslation(),
)(DeleteFeature);
