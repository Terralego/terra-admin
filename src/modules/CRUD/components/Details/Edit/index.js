import { withRouter } from 'react-router-dom';
import { withTranslation } from 'react-i18next';
import { connectAuthProvider } from '@terralego/core/modules/Auth';

import { connectAppProvider } from '../../../../../components/AppProvider';
import { connectCRUDProvider } from '../../../services/CRUDProvider';
import { connectMapProvider } from '../../../views/Map/MapProvider';
import { withTableFilters } from '../../../services/UserSettingsProvider';
import { getView, getLayersPaints } from '../../../services/CRUD';
import compose from '../../../../../utils/compose';

import Edit from './Edit';

const appProviderGetter = ({
  env: { modules: { CRUD: { settings } } },
}) => ({
  settingsEndpoint: settings,
});

const authProviderGetter = ({
  authenticated,
  user,
}) => {
  const permissions = authenticated ? user.permissions : [];
  return {
    displayAddFeature: permissions.includes('can_add_feature'),
    displayChangeFeature: permissions.includes('can_change_feature'),
  };
};

const CRUDPRoviderGetter = ({
  getSettings,
  getFeaturesList,
  settings,
  map,
  feature,
  saveFeature,
  errors: {
    feature: featureError,
  },
}, {
  match: { params: { layer, id } },
}) => ({
  getSettings,
  getFeaturesList,
  map,
  feature: feature[id] || {},
  saveFeature,
  view: getView(settings, layer),
  layerPaint: getLayersPaints(settings).find(item => item['source-layer'] === layer) || {},
  paramLayer: layer,
  paramId: id,
  featureError: featureError.find(({ featureId }) => featureId === id),
});

export default compose(
  withRouter,
  connectAppProvider(appProviderGetter),
  connectAuthProvider(authProviderGetter),
  connectCRUDProvider(CRUDPRoviderGetter),
  connectMapProvider('addControl', 'removeControl'),
  withTableFilters(),
  withTranslation(),
)(Edit);
