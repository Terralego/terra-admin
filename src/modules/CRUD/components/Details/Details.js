import React from 'react';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';

import { ACTION_CREATE } from '../../services/CRUD';
import Loading from '../../../../components/Loading';
import { generateURI } from '../../config';
import { toast } from '../../../../utils/toast';

import Metas from './Metas';
import Read from './Read';
import Create from './Create';
import './styles.scss';

class Details extends React.Component {
  static propTypes = {
    view: PropTypes.shape({
      featureEndpoint: PropTypes.string,
    }),
    feature: PropTypes.shape({
      properties: PropTypes.shape({}),
    }),
    fetchFeature: PropTypes.func.isRequired,
    match: PropTypes.shape({
      params: PropTypes.shape({
        id: PropTypes.string,
        layer: PropTypes.string,
        action: PropTypes.string,
      }),
    }),
    hasError: PropTypes.bool,
    errorMessage: PropTypes.string,
    detailsHasLoaded: PropTypes.func,
    onSizeChange: PropTypes.func,
    t: PropTypes.func,
  };

  static defaultProps = {
    view: {
      featureEndpoint: undefined,
    },
    feature: {
      properties: {},
    },
    match: {
      params: {
        id: undefined,
        layer: undefined,
        action: undefined,
      },
    },
    hasError: false,
    errorMessage: '',
    detailsHasLoaded () {},
    onSizeChange () {},
    t: text => text,
  }

  detailContent = React.createRef();

  componentDidMount () {
    const {
      match: { params: { id: paramId } },
      detailsHasLoaded,
    } = this.props;
    this.getData();
    if (paramId === ACTION_CREATE) {
      detailsHasLoaded();
    }
  }

  componentDidUpdate ({
    match: {
      params: {
        id: prevParamId,
        layer: prevParamlayer,
      },
    },
    feature: prevFeature,
  }) {
    const {
      match: {
        params: {
          id: paramId,
          layer: paramLayer,
        },
      },
      feature,
      detailsHasLoaded,
    } = this.props;

    if (feature !== prevFeature) {
      detailsHasLoaded();
    }

    if (prevParamlayer !== paramLayer || prevParamId !== paramId) {
      this.getData();
    }
  }

  get isCreateAction () {
    const { match: { params: { id: paramId } } } = this.props;
    return paramId === ACTION_CREATE;
  }

  getData () {
    const {
      view,
      fetchFeature,
      match: { params: { id: paramId } },
    } = this.props;
    if (view && paramId && !this.isCreateAction) {
      const { featureEndpoint } = view;
      fetchFeature(featureEndpoint, paramId);
    }
  }

  renderContent = () => {
    const {
      match: { params: { id } },
      feature,
    } = this.props;

    if (id === ACTION_CREATE) {
      return <Create />;
    }

    return (
      <Read feature={feature} />
    );
  }

  onSizeChange = () => {
    const { full, onSizeChange } = this.props;
    onSizeChange({ full: !full });
  }

  render () {
    const {
      feature,
      match: { params: { id: paramId, layer: paramLayer } },
      t,
      hasError,
      errorMessage,
      full,
    } = this.props;

    if (hasError && errorMessage === 'Not Found') {
      toast.displayError(t('CRUD.details.errorNoFeature'));
      return <Redirect to={generateURI('layer', { layer: paramLayer })} />;
    }
    const isLoading = !Object.keys(feature).length && paramId !== ACTION_CREATE;

    return (
      <>
        <Metas
          full={full}
          onSizeChange={this.onSizeChange}
        />
        <div ref={this.detailContent} className="CRUD-details__content">
          {isLoading
            ? <Loading spinner />
            : this.renderContent()}
        </div>
      </>
    );
  }
}

export default Details;
