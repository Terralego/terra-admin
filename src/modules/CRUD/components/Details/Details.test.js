import React from 'react';
import renderer from 'react-test-renderer';
import { ACTION_CREATE } from '../../services/CRUD';

import Details from './Details';

jest.mock('react-router-dom', () => ({
  Redirect: () => <div>Error because Redirect</div>,
}));

jest.mock('../../services/CRUD', () => ({
  ACTION_CREATE: 'create',
}));

jest.mock('../../../../components/Loading', () => () => <div>Loading</div>);

jest.mock('../../../../utils/toast', () => ({
  toast: {
    displayError: jest.fn(),
  },
}));

jest.mock('./Metas', () => () => <div>Metas</div>);
jest.mock('./Create', () => () => <div>Create</div>);
jest.mock('./Read', () => () => <div>Read</div>);

let props;
beforeEach(() => {
  props = {
    fetchFeature: jest.fn(),
    t: key => key,
    view: {
      layer: {
        id: 8,
        name: 'fooLayer',
      },
      formSchema: {
        properties: {
          city: { type: 'boolean', title: 'City' },
        },
      },
    },
    feature: {
      fooFeature: {
        identifier: 'fooFeature',
        layer: 8,
        geom: {
          type: 'Point',
          coordinates: [
            0.334734168777004,
            48.593660025881405,
          ],
        },
        properties: {
          name: 'dfdf',
        },
      },
    },
  };
});

describe('Snapshots', () => {
  it('should redirect if there is an error', () => {
    const tree = renderer.create((
      <Details
        {...props}
        hasError
        errorMessage="Not Found"
      />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render loading correctly', () => {
    const tree = renderer.create((
      <Details
        {...props}
        feature={{}}
      />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render details Read correctly', () => {
    const tree = renderer.create((
      <Details
        {...props}
      />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render details Create correctly', () => {
    const tree = renderer.create((
      <Details
        {...props}
        match={{
          params: {
            id: ACTION_CREATE,
            layer: 'fooLayer',
          },
        }}
      />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

it('should order that details component has loaded when changing feature', () => {
  const match =  {
    params: {
      id: 'fooId',
      layer: 'fooLayer',
    },
  };
  const detailsHasLoaded =  jest.fn();
  const instance = new Details({
    ...props,
    match,
    detailsHasLoaded,
  });
  instance.componentDidUpdate({ ...props, match, feature: {} }, {});
  expect(detailsHasLoaded).toHaveBeenCalled();
});

it('should get data when changing param layer or id', () => {
  const match1 =  {
    params: {
      id: 'id1',
      layer: 'layer1',
    },
  };
  const match2 =  {
    params: {
      id: 'id2',
      layer: 'layer1',
    },
  };
  const match3 =  {
    params: {
      id: 'id2',
      layer: 'layer2',
    },
  };
  const instance = new Details({
    ...props,
    match: match1,
  });
  instance.setState = jest.fn();
  instance.getData = jest.fn();
  instance.componentDidUpdate({ ...props, match: match2 }, {});
  instance.componentDidUpdate({ ...props, match: match3 }, {});

  expect(instance.getData).toBeCalledTimes(2);
});

it('should change size', () => {
  const instance = new Details({
    ...props,
    match: {
      params: {
        id: 'fooFeature',
        layer: 'layer1',
      },
    },
    full: true,
    onSizeChange: jest.fn(),
  });
  instance.setState = jest.fn();
  instance.onSizeChange();

  expect(instance.props.onSizeChange).toHaveBeenCalledWith({
    full: false,
  });
});
