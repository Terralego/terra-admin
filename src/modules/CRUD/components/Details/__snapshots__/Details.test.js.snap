// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots should redirect if there is an error 1`] = `
<div>
  Error because Redirect
</div>
`;

exports[`Snapshots should render details Create correctly 1`] = `
Array [
  <div>
    Metas
  </div>,
  <div
    className="CRUD-details__content"
  >
    <div>
      Create
    </div>
  </div>,
]
`;

exports[`Snapshots should render details Read correctly 1`] = `
Array [
  <div>
    Metas
  </div>,
  <div
    className="CRUD-details__content"
  >
    <div>
      Read
    </div>
  </div>,
]
`;

exports[`Snapshots should render loading correctly 1`] = `
Array [
  <div>
    Metas
  </div>,
  <div
    className="CRUD-details__content"
  >
    <div>
      Loading
    </div>
  </div>,
]
`;
